source/opt/split_invalid_unreachable_pass.cpp

```diff
@@ -0,0 +1,87 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/split_invalid_unreachable_pass.h"
+
+#include "source/opt/ir_builder.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status SplitInvalidUnreachablePass::Process() {
+  bool changed = false;
+  std::unordered_set<uint32_t> entry_points;
+  for (auto entry_point : context()->module()->entry_points()) {
+    entry_points.insert(entry_point.GetSingleWordOperand(1));
+  }
+
+  for (auto func = context()->module()->begin();
+       func != context()->module()->end(); ++func) {
+    if (entry_points.find(func->result_id()) == entry_points.end()) continue;
+    std::unordered_set<uint32_t> continue_targets;
+    std::unordered_set<uint32_t> merge_blocks;
+    std::unordered_set<BasicBlock*> unreachable_blocks;
+    for (auto block = func->begin(); block != func->end(); ++block) {
+      unreachable_blocks.insert(&*block);
+      uint32_t continue_target = block->ContinueBlockIdIfAny();
+      if (continue_target != 0) continue_targets.insert(continue_target);
+      uint32_t merge_block = block->MergeBlockIdIfAny();
+      if (merge_block != 0) merge_blocks.insert(merge_block);
+    }
+
+    cfg()->ForEachBlockInPostOrder(
+        func->entry().get(), [&unreachable_blocks](BasicBlock* inner_block) {
+          unreachable_blocks.erase(inner_block);
+        });
+
+    for (auto unreachable : unreachable_blocks) {
+      uint32_t block_id = unreachable->id();
+      if (continue_targets.find(block_id) == continue_targets.end() ||
+          merge_blocks.find(block_id) == merge_blocks.end()) {
+        continue;
+      }
+      context()->get_def_use_mgr()->ForEachUse(
+          unreachable->GetLabelInst(),
+          [this, func, &changed](Instruction* use, uint32_t idx) {
+            if ((use->opcode() == SpvOpLoopMerge && idx == 0) ||
+                use->opcode() == SpvOpSelectionMerge) {
+              uint32_t new_id = context()->TakeNextId();
+              std::unique_ptr<Instruction> new_label(
+                  new Instruction(this->context(), SpvOpLabel, 0, new_id, {}));
+              get_def_use_mgr()->AnalyzeInstDefUse(new_label.get());
+              std::unique_ptr<BasicBlock> new_block(
+                  new BasicBlock(std::move(new_label)));
+              auto* block_ptr = new_block.get();
+              InstructionBuilder builder(
+                  context(), new_block.get(),
+                  IRContext::kAnalysisDefUse |
+                      IRContext::kAnalysisInstrToBlockMapping);
+              builder.AddUnreachable();
+              cfg()->RegisterBlock(block_ptr);
+              (&*func)->AddBasicBlock(std::move(new_block));
```

Comment:
It will look awkward when people look at the spir-v because the merge block will be at the end of the function.  It is not necessary, but it might be best to insert the new block before `unreachable`.

Reply:
Done